Refactor the authorization task

yangminzhu opened the pull request:
- Move the permissive mode to a standalone task
- Rename the group/list claim support to align with other tasks
- Re-order to put the basic HTTP/TCP task first

/cc @lei-tang 

Signed-off-by: Yangmin Zhu <ymzhu@google.com>

istio-testing commented:
@yangminzhu: GitHub didn't allow me to request PR reviews from the following users: lei-tang.

Note that only [istio members](https://github.com/orgs/istio/people) and repo collaborators can review this PR, and authors cannot review their own PRs.

<details>

In response to [this](https://github.com/istio/istio.io/pull/3150):

>- Move the permissive mode to a standalone task
>- Rename the group/list claim support to align with other tasks
>- Re-order to put the basic HTTP/TCP task first
>
>/cc @lei-tang 
>
>Signed-off-by: Yangmin Zhu <ymzhu@google.com>


Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

istio-testing commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/istio/istio.io/pull/3150#" title="Author self-approved">yangminzhu</a>*
To fully approve this pull request, please assign additional approvers.
We suggest the following additional approver: **sdake**

If they are not already assigned, you can assign the PR to them by writing `/assign @sdake` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details open>
Needs approval from an approver in each of these files:

- **[content/OWNERS](https://github.com/istio/istio.io/blob/master/content/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["sdake"]} -->

istio-testing commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/istio/istio.io/pull/3150#pullrequestreview-197249924" title="LGTM">liminw</a>*, *<a href="https://github.com/istio/istio.io/pull/3150#" title="Author self-approved">yangminzhu</a>*
To fully approve this pull request, please assign additional approvers.
We suggest the following additional approver: **sdake**

If they are not already assigned, you can assign the PR to them by writing `/assign @sdake` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details open>
Needs approval from an approver in each of these files:

- **[content/OWNERS](https://github.com/istio/istio.io/blob/master/content/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["sdake"]} -->

yangminzhu commented:
@geeknoid Could you review this? Thank you.

geeknoid commented:
I'd prefer if @rcaballeromx takes a look at this, since he's very familiar with the security content.

rcaballeromx commented:
I'm starting the review.
/hold

rcaballeromx commented on content/docs/tasks/security/authz-permissive/index.md:
```diff
@@ -0,0 +1,231 @@
+    The `permissiveResponseCode` is `denied` which is what user will see after switching
+    global authorization configuration from `PERMISSIVE` mode to `ENFORCED` mode, which
+    indicates the global authorization configuration will work as expected after rolling
+    to production.
```
This text fails to convey to readers what actions they should be taking. To the readers, the log is in front of them and they are seeing the log at that moment, thus the use of the future tense is confusing. The paragraph includes information that is not pertinent to the action the reader needs to take. Remove the unrelated information to make the text easier to follow. You do mention how the log changes later on, there is no need to obfuscate the action with that information at this point.
```suggestion
1. Verify that the the log shows a `responseCode` of `200` and a `permissiveResponseCode` of `denied`.
```

rcaballeromx commented on content/docs/tasks/security/authz-permissive/index.md:
```diff
@@ -0,0 +1,231 @@
+
+1.  Before rolling out a new authorization policy in production, apply it in permissive mode.
+    `Note`, when global authorization configuration is in permissive mode, all policies will be in
+    permissive mode by default.
```
Is the reader running the commands in this doc applying the policy before rolling out into production? The answer is clearly yes, thus the text creates more questions than it answers.  Readers might wonder whether or not they already did something wrong. If you consider it warranted, add it as a proper warning. The text explaining what the permissive mode is and how it works should not be part of the steps. See my previous comment on that regard. Add instead the information regarding what specific authorization policy you are talking about.
```suggestion
1. Apply the `productpage-policy.yaml` authorization policy in permissive mode with the following command:
```

rcaballeromx commented on content/docs/tasks/security/authz-permissive/index.md:
```diff
@@ -0,0 +1,231 @@
+    `Note`, when global authorization configuration is in permissive mode, all policies will be in
+    permissive mode by default.
+
+    Run the following command:
```
```suggestion
```

rcaballeromx commented on content/docs/tasks/security/authz-permissive/index.md:
```diff
@@ -0,0 +1,231 @@
+    $ kubectl apply -f @samples/bookinfo/platform/kube/rbac/productpage-policy.yaml@
+    {{< /text >}}
+
+1.  Send traffic to the sample application again.
```
Just copy the previous instance of this step. Avoid using "again" as it gives readers the impression that they are not making any progress.

rcaballeromx commented on content/docs/tasks/security/authz-permissive/index.md:
```diff
@@ -0,0 +1,231 @@
+    Point your browser at the Bookinfo `productpage` (`http://$GATEWAY_URL/productpage`), you should
+    see everything works fine.
+
+1.  Verify the logs and check `permissiveResponseCode` again.
```
See my previous comments on how to handle steps involving the logs.

rcaballeromx commented on content/docs/tasks/security/authz-permissive/index.md:
```diff
@@ -0,0 +1,231 @@
+    policy mode from `PERMISSIVE` mode to `ENFORCED` mode; the result aligns with
+    [Emabling authorization for HTTP services step 1](/docs/tasks/security/authz-http#step-1-allowing-access-to-the-productpage-service).
+
+1.  Remove permissive mode related yaml files:
```
The order of the words here can be confusing. Are readers supposed to remove the permissive mode or the files? There is a bit of context which is missing, why should readers remove those files?
```suggestion
1.  Remove the YAML files related to enabling the permissive mode:
```

rcaballeromx commented on content/docs/tasks/security/authz-permissive/index.md:
```diff
@@ -0,0 +1,231 @@
+    $ kubectl delete -f @samples/bookinfo/platform/kube/rbac/rbac-permissive-telemetry.yaml@
+    {{< /text >}}
+
+1.  Now we have verified authorization will work as expected when turning it on,
```
Address the reader directly and convey a sense of accomplishment before directing them to the next task. Be explicit about what exactly the reader has accomplished and about the next steps you recommend them taking. Your text doesn't do this very effectively consider instead:
```suggestion
1.  Congratulations! You tested an authorization policy and verified it will work as expected. To enable the authorization policy, follow the steps described in the [Enabling Istio authorization task](/docs/tasks/security/authz-http#enabling-istio-authorization).
```

rcaballeromx commented on content/docs/tasks/security/authz-permissive/index.md:
```diff
@@ -0,0 +1,231 @@
+1.  Now we have verified authorization will work as expected when turning it on,
+    it's safe following the [Enabling Istio authorization](/docs/tasks/security/authz-http#enabling-istio-authorization) to turn on authorization.
+
+### Testing new authorization policy works as expected before rolling to production
```
Similarly to the previous heading, this one is too long and also fails to convey the content of the section that follows.
```suggestion
### Testing an authorization policy with authentication enabled
```

yangminzhu replied:
Changed to `Test adding authorization policy`

rcaballeromx commented on content/docs/tasks/security/authz-permissive/index.md:
```diff
@@ -0,0 +1,231 @@
+### Testing new authorization policy works as expected before rolling to production
+
+This tasks shows how to use authorization permissive mode to test a new authorization policy works
+as expected in environment with authorization already enabled.
```
This text is confusing because of the grammar issues it has.
```suggestion
The following steps show how to test a new authorization policy when authorization has already been enabled.
```

rcaballeromx commented on content/docs/tasks/security/authz-permissive/index.md:
```diff
@@ -0,0 +1,231 @@
+This tasks shows how to use authorization permissive mode to test a new authorization policy works
+as expected in environment with authorization already enabled.
+
+Before you start, please make sure that you have finished [Enabling authorization for HTTP services step 1](/docs/tasks/security/authz-http#step-1-allowing-access-to-the-productpage-service).
```
I would strongly encourage you to replace this text with a step and a command to verify whether authentication is enabled.
```suggestion
1. Confirm that authentication is enabled with the following command:
```

rcaballeromx commented on content/docs/tasks/security/authz-permissive/index.md:
```diff
@@ -0,0 +1,231 @@
+
+Before you start, please make sure that you have finished [Enabling authorization for HTTP services step 1](/docs/tasks/security/authz-http#step-1-allowing-access-to-the-productpage-service).
+
+1.  Before applying a new policy, test it by setting its mode to permissive:
```
Please split this step as I have described in previous comments. Every step should contain a single action. This one has three: 1. Run the command, 2. Review the YAML, and 3. Verify there is an error when going to the product page.

rcaballeromx commented on content/docs/tasks/security/authz-permissive/index.md:
```diff
@@ -0,0 +1,231 @@
+    product reviews` on the page. These errors are expected because the policy is
+    in `PERMISSIVE` mode.
+
+1.  Apply YAML file for the permissive mode metric collection.
```
Be specific about what file you mean, there are many YAML files in Istio.
```suggestion
1.  Apply the `rbac-permissive-telemetry.yaml` YAML file to enable the permissive mode metric collection.
```

rcaballeromx commented on content/docs/tasks/security/authz-permissive/index.md:
```diff
@@ -0,0 +1,231 @@
+
+1.  Apply YAML file for the permissive mode metric collection.
+
+    Run the following command:
```
```suggestion
```

rcaballeromx commented on content/docs/tasks/security/authz-permissive/index.md:
```diff
@@ -0,0 +1,231 @@
+    $ kubectl apply -f @samples/bookinfo/platform/kube/rbac/rbac-permissive-telemetry.yaml@
+    {{< /text >}}
+
+1.  Send traffic to the sample application.
```
Use the same step as in the previous section.

rcaballeromx commented on content/docs/tasks/security/authz-permissive/index.md:
```diff
@@ -0,0 +1,231 @@
+    $ curl http://$GATEWAY_URL/productpage
+    {{< /text >}}
+
+1.  Verify the logs and check `permissiveResponseCode` again.
```
See my previous comments regarding the steps needed for logs.

rcaballeromx commented on content/docs/tasks/security/authz-permissive/index.md:
```diff
@@ -0,0 +1,231 @@
+    to `ENFORCED` mode; it indicates the new authorization policy will work as expected
+    after rolling to production.
+
+1.  Remove permissive mode related yaml files:
```
See my previous comment on the word order.

rcaballeromx commented on content/docs/tasks/security/authz-permissive/index.md:
```diff
@@ -0,0 +1,231 @@
+    $ kubectl delete -f @samples/bookinfo/platform/kube/rbac/rbac-permissive-telemetry.yaml@
+    {{< /text >}}
+
+1.  Now we have verified the new policy will work as expected, it's safe
```
Please use the same text I suggested at the end of the previous section.

rcaballeromx commented:
/hold cancel

yangminzhu commented on content/docs/tasks/security/authz-permissive/index.md:
```diff
@@ -0,0 +1,231 @@
+keywords: [security,access-control,rbac,authorization]
+---
+
+The authorization permissive mode is an experimental feature in Istio's 1.1 release. Its interface can change in future releases.
```
I assume you're asking to change the `Istio's 1.1 release` to `1.1 version`, is that right?

yangminzhu commented on content/docs/tasks/security/authz-permissive/index.md:
```diff
@@ -0,0 +1,231 @@
+
+* Understand [authorization](/docs/concepts/security/#authorization) concepts.
+
+* Have set up Istio on Kubernetes **with authentication enabled** by following the instructions in the
```
Note It should be installing Istio with authentication, not Kubernetes with authentication.

yangminzhu commented on content/docs/tasks/security/authz-permissive/index.md:
```diff
@@ -0,0 +1,231 @@
+
+* Deploy the [Bookinfo](/docs/examples/bookinfo/) sample application.
+
+### Testing whether it's safe to turn on authorization globally
```
This is about authorization and has nothing to do with authentication. I changed to `Test enabling authorization globally`

yangminzhu commented on content/docs/tasks/security/authz-permissive/index.md:
```diff
@@ -0,0 +1,231 @@
+### Testing whether it's safe to turn on authorization globally
+
+This tasks show how to use authorization permissive mode to test whether it's safe to
+turn on authorization globally.
```
This section is about enabling authorization globally, next section is about test an specific authorization policy.

Changed to `The following steps show you how to use authorization permissive mode to test whether it's safe to
turn on authorization globally:`

yangminzhu commented on content/docs/tasks/security/authz-permissive/index.md:
```diff
@@ -0,0 +1,231 @@
+1.  Verify the logs stream has been created and check `permissiveResponseCode`.
+
+    In a Kubernetes environment, search through the `istio-telemetry`
+    pods' logs as follows:
```
Changed to `Get the log for telemetry and search for the `permissiveResponseCode` with the following command:`

istio-testing commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/istio/istio.io/pull/3150#pullrequestreview-197249924" title="LGTM">liminw</a>*, *<a href="https://github.com/istio/istio.io/pull/3150#" title="Author self-approved">yangminzhu</a>*
To fully approve this pull request, please assign additional approvers.
We suggest the following additional approver: **geeknoid**

If they are not already assigned, you can assign the PR to them by writing `/assign @geeknoid` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details open>
Needs approval from an approver in each of these files:

- **[content/OWNERS](https://github.com/istio/istio.io/blob/master/content/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["geeknoid"]} -->

yangminzhu commented:
@rcaballeromx 

I have addressed the comments so far, Please take a look, Thanks!

I copied all contents from existing docs, so based on your comments I also changed some of them in the original places (authz-http/index.md)

rcaballeromx commented on content/docs/tasks/security/authz-permissive/index.md:
```diff
@@ -0,0 +1,179 @@
+
+* Deploy the [Bookinfo](/docs/examples/bookinfo/) sample application.
+
+* Create service accounts and for the Bookinfo application. Run the following command to create service
```
```suggestion
* Create service accounts for the Bookinfo application. Run the following command to create service
```

rcaballeromx commented on content/docs/tasks/security/authz-permissive/index.md:
```diff
@@ -0,0 +1,179 @@
+    $ kubectl delete -f @samples/bookinfo/platform/kube/rbac/rbac-permissive-telemetry.yaml@
+    {{< /text >}}
+
+1.  Congratulations! You tested an authorization policy with permissive mode and verified it will
```
```suggestion
1.  Congratulations! You tested an authorization policy with permissive mode and verified it
```

rcaballeromx commented on content/docs/tasks/security/authz-permissive/index.md:
```diff
@@ -0,0 +1,179 @@
+    {{< /text >}}
+
+1.  Congratulations! You tested an authorization policy with permissive mode and verified it will
+    work as expected. To enable the authorization policy, follow the steps described in the
```
```suggestion
    works as expected. To enable the authorization policy, follow the steps described in the
```

rcaballeromx commented on content/docs/tasks/security/authz-permissive/index.md:
```diff
@@ -0,0 +1,179 @@
+The following steps show how to test a new authorization policy with permissive mode when authorization
+has already been enabled.
+
+1.  Allow access to the `producepage` service by following
```
```suggestion
1.  Allow access to the `producepage` service by following the instructions in
```

rcaballeromx commented on content/docs/tasks/security/authz-http/index.md:
```diff
@@ -16,22 +16,15 @@ a basic tutorial in Istio Security Basics.
-    * Create service account `bookinfo-productpage`, and redeploy the service `productpage` with the service account.
-    * Create service account `bookinfo-reviews`, and redeploy the services `reviews` (deployments `reviews-v2` and `reviews-v3`)
-    with the service account.
+* Create service accounts and for the Bookinfo application. Run the following command to create service
```
Typo
```suggestion
* Create service accounts for the Bookinfo application. Run the following command to create service
```

istio-testing commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/istio/istio.io/pull/3150#pullrequestreview-197249924" title="LGTM">liminw</a>*, *<a href="https://github.com/istio/istio.io/pull/3150#pullrequestreview-200295392" title="LGTM">rcaballeromx</a>*, *<a href="https://github.com/istio/istio.io/pull/3150#" title="Author self-approved">yangminzhu</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[content/OWNERS](https://github.com/istio/istio.io/blob/master/content/OWNERS)~~ [rcaballeromx]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->
